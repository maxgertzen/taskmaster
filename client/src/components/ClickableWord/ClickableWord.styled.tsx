import styled from '@emotion/styled';

export const StyledLink = styled('a')(({ theme }) => ({
  color: theme.colors.primary.accent,
  fontWeight: 500,
  textDecorationColor: theme.colors.primary.accent,
  cursor: 'pointer',

  '&:hover': {
    textDecoration: 'underline',
  },
}));
