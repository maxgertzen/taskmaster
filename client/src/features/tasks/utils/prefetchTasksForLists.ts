import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
