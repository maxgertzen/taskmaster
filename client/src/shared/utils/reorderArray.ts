export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
