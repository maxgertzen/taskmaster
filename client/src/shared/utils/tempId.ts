export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
