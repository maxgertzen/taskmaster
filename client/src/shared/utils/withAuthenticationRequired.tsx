import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
