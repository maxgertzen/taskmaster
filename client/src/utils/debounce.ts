export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
