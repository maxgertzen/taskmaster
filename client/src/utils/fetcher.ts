export type FetcherConfig<K> = {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: K;
  timeout?: number;
  headers?: HeadersInit;
  urlSearchParams?: string;
};

export const fetcher = async <T, K = undefined>(
  config: FetcherConfig<K>
): Promise<T> => {
  try {
    const {
      method = 'GET',
      body = {},
      headers = {},
      timeout = 5000,
      urlSearchParams,
    } = config;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      ...(method !== 'GET' && { body: JSON.stringify(body) }),
      signal: controller.signal,
    };

    if (urlSearchParams) {
      config.url += `?${urlSearchParams}`;
    }

    const response = await fetch(config.url, options);
    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Failed to fetch data:', error);
    throw error;
  }
};
