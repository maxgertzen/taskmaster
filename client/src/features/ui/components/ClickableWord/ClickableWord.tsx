import { FC, MouseEvent, ReactNode } from 'react';

import { useSpotlightStore } from '@/shared/store/spotlightStore';

import { StyledLink } from './ClickableWord.styled';

interface ClickableWordProps {
  target: string;
  children: ReactNode;
}

export const ClickableWord: FC<ClickableWordProps> = ({ target, children }) => {
  const setTarget = useSpotlightStore((state) => state.setTarget);

  const handleClick = (event: MouseEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    setTarget(target);
    setTimeout(() => {
      setTarget(null);
    }, 1500);
  };
  return (
    <StyledLink href='#' onClick={handleClick}>
      {children}
    </StyledLink>
  );
};
