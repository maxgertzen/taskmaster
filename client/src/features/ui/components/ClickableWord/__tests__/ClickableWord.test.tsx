import { describe, it, expect, vi, beforeEach, afterAll } from 'vitest';

import { forceClick, render, screen } from '@/tests/utils';

import { ClickableWord } from '../ClickableWord';

type SpotlightStore = {
  setTarget: (target: string | null) => void;
};

const mockSetTarget = vi.fn();
vi.mock('@/shared/store/spotlightStore', () => ({
  useSpotlightStore: (selector: (state: SpotlightStore) => unknown) =>
    selector({ setTarget: mockSetTarget }),
}));

describe('ClickableWord', () => {
  beforeEach(() => {
    mockSetTarget.mockClear();
  });

  afterAll(() => {
    vi.useRealTimers();
  });

  it('renders correctly in light theme', () => {
    const { container } = render(
      <ClickableWord target='test-target'>Click me</ClickableWord>,
      { theme: 'light' }
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correctly in dark theme', () => {
    const { container } = render(
      <ClickableWord target='test-target'>Click me</ClickableWord>,
      { theme: 'dark' }
    );
    expect(container).toMatchSnapshot();
  });

  it('renders the children content', () => {
    render(<ClickableWord target='test-target'>Test Content</ClickableWord>);
    const link = screen.getByTestId('clickable-word');
    expect(link.textContent).toBe('Test Content');
  });

  it('calls setTarget with correct target when clicked', async () => {
    render(<ClickableWord target='test-target'>Click me</ClickableWord>);

    const link = screen.getByTestId('clickable-word');
    const { preventDefaultMock } = forceClick(link);

    expect(preventDefaultMock).toHaveBeenCalled();
    expect(mockSetTarget).toHaveBeenCalledWith('test-target');
  });

  it('resets target to null after timeout', async () => {
    vi.useFakeTimers();
    render(<ClickableWord target='test-target'>Click me</ClickableWord>);
    const link = screen.getByTestId('clickable-word');

    forceClick(link);

    expect(mockSetTarget).toHaveBeenCalledWith('test-target');
    await vi.advanceTimersByTimeAsync(1500);
    expect(mockSetTarget).toHaveBeenLastCalledWith(null);
  });
});
