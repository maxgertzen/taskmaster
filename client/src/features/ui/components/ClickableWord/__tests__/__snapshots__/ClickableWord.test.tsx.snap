// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ClickableWord > renders correctly in dark theme 1`] = `
<div>
  <a
    class="css-t1i85f"
    data-testid="clickable-word"
    href="#"
  >
    Click me
  </a>
</div>
`;

exports[`ClickableWord > renders correctly in light theme 1`] = `
<div>
  <a
    class="css-1hmk8d3"
    data-testid="clickable-word"
    href="#"
  >
    Click me
  </a>
</div>
`;
